guides/v2.0/coding-standards/technical-guidelines.md

```diff
@@ -525,7 +527,81 @@ class View extends Template
+
+6.4.2.3. Each service interface SHOULD declare a single public method. An interface name SHOULD reflect the task or action to be performed. For example, `Magento\InventoryApi\Api\StockSourceLinksDeleteInterface::execute(array $links)`. The only exception is a Repository API, which MAY be added for convenience and MUST be limited to singular CRUD operations and `getList($searchCriteria)`.
+
+6.4.3. Service Method Signature
+
+6.4.3.1. Strict typing is enforced for Service and Data interfaces located under `MyCompany/MyModuleApi/Api`. Only the following types are allowed:
+
+* Scalar types: `string` (including Date and DateTime); `int`; `float`; `boolean`
+
+* Data interfaces
+
+* One-dimensional indexed arrays of scalars or data interfaces: for example `string[]`, `\MyCompany\MyModuleApi\Api\Data\SomeInterface[]`. Hash maps (associative arrays) are not supported.
+
+* Nullable scalars or data interfaces: for example `string|null`. Using just `null` is prohibited.
+
+* `void`
+
+6.4.3.2. Service contracts SHOULD support batch data processing. For example, an entity persisting method SHOULD accept an array of entities to persist instead of a single entity. Customizations implemented through plugins SHOULD be adjusted respectively. 
+
+6.4.3.3. Batch retrieval operations MUST accept `SearchCriteriaInterface` and return `SearchResultInterface` to support pagination.
+
+6.4.3.4. Batch operations that modify state MUST accept an array of entities and return a response object that contains:
+
+* An array of successfully processed items
+
+* An array of items with retriable errors
+
+* An array of items with non-retriable errors
+
+6.4.3.5. Batch operations that modify state SHOULD be implemented in the most performant manner and SHOULD NOT load modified entities to generate response. 
+
+6.4.3.6. Asynchronous invocation of the command services SHOULD be supported by the web API framework.
+
+6.4.3.7. Operation UUID MAY be provided by the client during service invocation. UUID MUST allow the client to get the operation status information.
+
+6.4.3.8. Data objects returned by service contracts SHOULD be fully loaded to ensure consistency.
+
+6.4.4. Service Implementation
+
+6.4.4.1. Service data interfaces SHOULD extend from `Magento\Framework\Api\ExtensibleDataInterface`. The only exception is when extensibility is not desired, such as in case of value-objects.
+
+6.4.4.2. Extensible data interfaces MUST NOT form hierarchies. If interface `MyInterface` extends `ExtensibleDataInterface`, there must be no interfaces extending `MyInterface`. Otherwise, a list of extension attributes will be shared for all extensible interfaces in the hierarchy.
+
+6.4.4.3. Service implementations and plugins MUST NOT rely on storage-specific integrity features, such as foreign key constraints.
+
+6.4.4.4. Replacement strategy SHOULD be used to persist main entity fields/attributes, child entities, and relation links. 
```

Comment:
6.4.4.4 and 6.4.4.5 better replace with:  Persist of main entity fields/attributes, child entities, and relation links MUST be *idempotent*

Reply:
I would keep 6.4.4.5, 
6.4.4.4 can be changed to have `idempotency` requirement